### PR TITLE
Generalize Intel coalescing pass to handle users of `scf.for` with coalesced load

### PR DESCRIPTION
**Pull request by etiotto**
Ensure that loads of a `scf.for` yielded value with block ptr type can be coalesced

### CODE REVIEW

File: test/TritonIntelGPU/coalesce.mlir [hunk 1]
```diff
@@ -382,4 +382,35 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 16
     }) : (tensor<32x128xf32, #blocked>) -> tensor<32xf32, #triton_gpu.slice<{dim = 1, parent = #blocked}>>
     tt.return
   }
+
+  // CHECK: @issue_2762
+  tt.func public @issue_2762(%arg0: !tt.ptr<f32> {tt.divisibility = 16 : i32}) {
```
Comment (whitneywhtsang) [inline test/TritonIntelGPU/coalesce.mlir:387]: There is room of simplification of the lit test.

Comment (whitneywhtsang) [reply test/TritonIntelGPU/coalesce.mlir:387]: e.g., I suspect the load in the loop is not needed to show the problem.
operations before the make_tensor_ptr not needed?

Comment (etiotto) [reply test/TritonIntelGPU/coalesce.mlir:387]: ok I am reducing the test a bit

File: test/TritonIntelGPU/coalesce.mlir [hunk 2]
```diff
@@ -382,4 +382,30 @@ module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 16
...
+    %5 = arith.remsi %1, %c512_i32 : i32
+    // CHECK: [[PTR1:%.*]] = tt.make_tensor_ptr %arg0, {{.*}} : <tensor<1x32x128xf32, [[BLOCKED_LAYOUT1]]>>
+    %y = tt.make_tensor_ptr %arg0, [%c512_i64, %c512_i64, %c512_i64], [%c1_i64, %c512_i64, %c262144_i64], [%4, %5, %c0_i32] {order = array<i32: 2, 1, 0>} : <tensor<1x32x128xf32, #blocked1>>
+    // CHECK: [[RES:%.*]] = scf.for {{.*}} iter_args([[ARG1:%.*]] = [[PTR1]]) -> (!tt.ptr<tensor<1x32x128xf32, [[BLOCKED_LAYOUT1]]>>)
+    %8:1 = scf.for %arg5 = %c0_i32 to %c512_i32 step %c128_i32 iter_args(%arg7 = %y) -> (!tt.ptr<tensor<1x32x128xf32, #blocked1>>) : i32 {
+      // CHECK: scf.yield [[ARG1]] : !tt.ptr<tensor<1x32x128xf32, [[BLOCKED_LAYOUT1]]>>
+      scf.yield %arg7 : !tt.ptr<tensor<1x32x128xf32, #blocked1>>
+    }
+    // CHECK: [[LOAD_RES:%.*]] = tt.load [[RES]] : !tt.ptr<tensor<1x32x128xf32, [[BLOCKED_LAYOUT1]]>>
+    // CHECK: triton_gpu.convert_layout [[LOAD_RES]] : tensor<1x32x128xf32, [[BLOCKED_LAYOUT1]]> -> tensor<1x32x128xf32, [[BLOCKED_LAYOUT2]]>
```
Comment (whitneywhtsang) [inline test/TritonIntelGPU/coalesce.mlir:407]:
```suggestion
    // CHECK: ttg.convert_layout [[LOAD_RES]] : tensor<1x32x128xf32, [[BLOCKED_LAYOUT1]]> -> tensor<1x32x128xf32, [[BLOCKED_LAYOUT2]]>
```

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**